### PR TITLE
Add description for SIONlib file format to backend documentation

### PR DESCRIPTION
**Pull request by jougs**

### CODE REVIEW

File: nestkernel/recording_backend_sionlib.h [hunk 1]
```diff
@@ -37,15 +37,19 @@ Store data to an efficient binary format
...
+<configure-for-parallel-computing>`.
+
+This backend is especially useful for large-scale simulations running
+in a distributed way on many MPI processes/OpenMP threads. In such
+usage scenarios, writing to plain text files (see :ref:`ASCII backend
 <ascii_backend>`) would cause a large overhead because of the huge
-number of generated files and be very inefficient. For the
+number of generated files and thus be very inefficient. For the
 implementation of writing to binary container files, NEST relies on
 the SIONlib library (http://www.fz-juelich.de/jsc/sionlib). Depending
```
Comment (sarakonradi) [inline nestkernel/recording_backend_sionlib.h:51]:
```suggestion
the `SIONlib library <http://www.fz-juelich.de/jsc/sionlib>`_. Depending
```
```suggestion
the SIONlib library (http://www.fz-juelich.de/jsc/sionlib). Depending
```

File: nestkernel/recording_backend_sionlib.h [hunk 2]
```diff
@@ -75,12 +79,39 @@ filename is composed.
...
+   NEST SIONlib binary file format.
+
+
+Reading the data
+++++++++++++++++
+
+As the binary format of the files produced by the ``sionlib`` backend,
+we provide a custom reader module for Python that makes the files
+available in a convenient way. The source code for this can be found
+in its own repository at https://github.com/nest/nest-sionlib-reader.
```
Comment (sarakonradi) [inline nestkernel/recording_backend_sionlib.h:114]:
```suggestion
in its own `repository <https://github.com/nest/nest-sionlib-reader>`_.
```
```suggestion
in its own repository at https://github.com/nest/nest-sionlib-reader.
```

**Review by sarakonradi — CHANGES_REQUESTED** [review]
Thanks for adding this description! Just added two minor suggestions regarding the links.

In which index is ``recording_backend_sionlib.h`` supposed to appear? I was able to find this page only through "search docs".

File: nestkernel/recording_backend_sionlib.h [hunk 3]
```diff
@@ -75,12 +79,39 @@ filename is composed.
...
+  individual record depends on the type of the device and is described
+  by a corresponding entry in the *device info block*
+  
+* the *file info block* keeps the file's metadata like version
+  information and such.
+
+* the *device info block* stores the properties and a data layout
+  description for each device that uses the ``sionlib`` backend.
+
+* the *tail block* contains pointers to the *file info block*
```
Comment (hakonsbm) [inline nestkernel/recording_backend_sionlib.h:97]: Two of the four bullet points here end with a period. Also, as they are all full sentences, they should start with a capital letter. Periods are not absolutely essential, but it should be consistent.

File: nestkernel/recording_backend_sionlib.h [hunk 4]
```diff
@@ -75,12 +79,39 @@ filename is composed.
...
+
+   NEST SIONlib binary file format.
+
+
+Reading the data
+++++++++++++++++
+
+As the binary format of the files produced by the ``sionlib`` backend,
+we provide a custom reader module for Python that makes the files
+available in a convenient way. The source code for this can be found
```
Comment (hakonsbm) [inline nestkernel/recording_backend_sionlib.h:113]: Is there something missing in the first part of this sentence? Right now it seems to imply that we are the binary format, which is a bit too existential for this documentation, I think.

**Review by hakonsbm — CHANGES_REQUESTED** [review]
Looks pretty good! I have a couple comments/suggestions.

**Comment by jougs** [comment]
Replying to @sarakonradi's comment:
> In which index is `recording_backend_sionlib.h` supposed to appear? I was able to find this page only through "search docs".

This is not supposed to be in any index, but the reStructuredText block is rather just included verbatim in the [Guide on recording from simulations](https://nest-simulator.readthedocs.io/en/latest/guides/recording_from_simulations.html).

File: nestkernel/recording_backend_sionlib.h (hunk 4 above)
Comment (jougs) [reply nestkernel/recording_backend_sionlib.h]: Fixed by 169047c47.

File: nestkernel/recording_backend_sionlib.h (hunk 3 above)
Comment (jougs) [reply nestkernel/recording_backend_sionlib.h]: Fixed by bcaae82f9.

File: nestkernel/recording_backend_sionlib.h [hunk 5]
```diff
@@ -75,12 +80,41 @@ filename is composed.
...
+  individual record depends on the type of the device and is described
+  by a corresponding entry in the *device info block*
+  
+* the *file info block* keeps the file's metadata like version
+  information and such
+
+* the *device info block* stores the properties and a data layout
+  description for each device that uses the ``sionlib`` backend
+
+* the *tail block* contains pointers to the *file info block*.
```
Comment (hakonsbm) [inline nestkernel/recording_backend_sionlib.h:98]: @jougs bcaae82f9a1a6d45477f0a232db6f9a65e394dc3 both removes periods, and adds a new one. so it's still inconsistent. Also, I still think the bullet points here should start with a capital letter (at least according to [this guide](https://www.lexico.com/grammar/bullet-points)).

```suggestion
* The *body block* contains the actual data records; the layout of an
  individual record depends on the type of the device and is described
  by a corresponding entry in the *device info block*
  
* The *file info block* keeps the file's metadata, like version
  information and such

* The *device info block* stores the properties and a data layout
  description for each device that uses the ``sionlib`` backend

* The *tail block* contains pointers to the *file info block*
```

**Review by hakonsbm — CHANGES_REQUESTED** [review]

File: nestkernel/recording_backend_sionlib.h (hunk 5 above)
Comment (jougs) [reply nestkernel/recording_backend_sionlib.h:98]: Agreed!

**Review by hakonsbm — APPROVED** [review]
Thanks! I'm happy now.

File: nestkernel/recording_backend_sionlib.h [hunk 6]
```diff
@@ -75,12 +80,41 @@ filename is composed.
...
+
+Reading the data
+++++++++++++++++
+
+As the binary format of the files produced by the ``sionlib`` does not
+conform to any standard, parsing them manually might be a bit
+cumbersome. To ease this task, we provide a reader module for Python
+that makes the files available in a convenient way. The source code
+and further documentation for this module can be found in its own
+repository at `<https://github.com/nest/nest-sionlib-reader>`_.
```
Comment (sarakonradi) [inline nestkernel/recording_backend_sionlib.h:117]:
```suggestion
`repository <https://github.com/nest/nest-sionlib-reader>`_.
```

File: nestkernel/recording_backend_ascii.h [hunk 7]
```diff
@@ -41,11 +41,11 @@ simulations, where the ease of a simple data format outweighs the
...
 
 This backend will open one file per recording device per thread on
-each MPI process. This can cause a high load on the file system
-in large simulations. This backend can become prohibitively inefficient,
-particularly on machines with distributed filesystems.
-In case you experience such scaling problems,  the :ref:`SIONlib
-backend <sionlib_backend>` may be a possible alternative.
+each MPI process. This can cause a high load on the file system in
+large simulations. This backend can become prohibitively inefficient,
+particularly on machines with distributed filesystems.  In case you
```
Comment (sarakonradi) [inline nestkernel/recording_backend_ascii.h:46]:
```suggestion
particularly on machines with distributed filesystems. In case you
```

**Review by sarakonradi — CHANGES_REQUESTED** [review]
Thank you for the changes! Please find two (very minor) comments below.

**Comment by sarakonradi** [comment]
I am happy, but Travis seems to complaining. Could you please check the formatting errors? Thanks!

**Comment by jougs** [comment]
Sorry for converting back to draft. I decided that I want to cross-link usage and compilation information similar to what I did #1830. Stay tuned!

**Comment by heplesser** [comment]
After re-running #5414.3 on Travis manually, it passed. I now started #5414.5 and 6, see https://travis-ci.org/github/nest/nest-simulator/builds/742795893. I am still a bit worried about why the test failed that once.

**Comment by jougs** [comment]
OK. I've revised the documentation in 96bba75 and think the code is now ready for review.

File: nestkernel/recording_backend_sionlib.h [hunk 8]
```diff
@@ -31,24 +31,31 @@
...
-the global settings of the `sionlib` recording backend (see below),
-either a single container file or a set of these files is created. In
-case of a single file, it is named according to the following pattern:
+distributed way on many MPI processes/OpenMP threads. In such usage
+scenarios, writing to plain text files (see :ref:`recording backend
+for ASCII files <recording_backend_ascii>`) would cause a large
+overhead because of the huge number of generated files and thus be
+very inefficient.
+
+The implementation of the `sionlib` backend is bases on the `SIONlib
```
Comment (hakonsbm) [inline nestkernel/recording_backend_sionlib.h:53]:
```suggestion
The implementation of the `sionlib` backend is based on the `SIONlib
```

**Review by hakonsbm — CHANGES_REQUESTED** [review]

**Review by sarakonradi — CHANGES_REQUESTED** [review]
Thanks for revising the documentation! I have added some comments inline.

File: doc/installation/compilation_options.rst [hunk 9]
```diff
@@ -113,29 +113,44 @@ by ``make install-nodoc`` to skip the generation of help pages and indices. Usin
...
-     try ``-Dwith-mpi=/path/to/my/mpi``.
-     Directory mpi should contain include, lib, bin subdirectories for MPI.
-  3. If that does not work, but you know the correct compiler wrapper for
-     your machine, try configure ``-DMPI_CXX_COMPILER=myC++_CompilerWrapper
+NEST supports distributed simulations using the Message Passing
+Interface (MPI). Depending on your setup, you have to use one of the
+following steps in order to add support for MPI:
+
+  1. Try ``-Dwith-mpi=ON`` as argument for ``cmake``. If this works,
+     fine.
```
[the remark this answers is not in the document]
Comment (jougs) [reply doc/installation/compilation_options.rst:126]: No. It was in the original text and I thought I leave it there, as someone might actually have liked it.

```suggestion
  1. Try ``-Dwith-mpi=ON`` as argument for ``cmake``.
```

**Comment by jougs** [comment]
Thanks for the repeated reviews (and your patience). I think now we're finally there :-)

**Review by hakonsbm — APPROVED** [review]
Great, thanks!

**Review by sarakonradi — APPROVED** [review]
LGTM, thanks for the repeated revisions :+1:

**Pull request merged**